test/e2e/framework/daemonset/fixtures.go

```diff
@@ -138,18 +140,20 @@ func checkDaemonPodStateOnNodes(ctx context.Context, c clientset.Interface, ds *
 	return len(nodesToPodCount) == len(nodeNames), nil
 }
 
-// CheckDaemonStatus returns false if not all desired pods are scheduled or not all of them are ready.
-func CheckDaemonStatus(ctx context.Context, f *framework.Framework, dsName string) func(ctx context.Context) (bool, error) {
-	return func(ctx context.Context) (bool, error) {
-		ds, err := f.ClientSet.AppsV1().DaemonSets(f.Namespace.Name).Get(ctx, dsName, metav1.GetOptions{})
-		if err != nil {
-			return false, err
-		}
-		desired, scheduled, ready := ds.Status.DesiredNumberScheduled, ds.Status.CurrentNumberScheduled, ds.Status.NumberReady
-		if desired == scheduled && scheduled == ready {
-			return true, nil
-		}
-		framework.Logf("error in daemon status. DesiredScheduled: %d, CurrentScheduled: %d, Ready: %d", desired, scheduled, ready)
-		return false, nil
-	}
+// CheckDaemonStatus ensures that eventually the daemon set has the desired
+// number of pods scheduled and ready. It returns a descriptive error if that
+// state is not reached in the amount of time it takes to start
+// pods. f.Timeouts.PodStart can be changed to influence that timeout.
+func CheckDaemonStatus(ctx context.Context, f *framework.Framework, dsName string) error {
+	return framework.Gomega().Eventually(ctx, framework.GetObject(f.ClientSet.AppsV1().DaemonSets(f.Namespace.Name).Get, dsName, metav1.GetOptions{})).
+		WithTimeout(f.Timeouts.PodStart).
```

Comment:
Nit: but I'd probably move the timeout information to the caller, so that the fact this function doesn't check just once but waits up to timeout is more readable. Probably something we'll want to fix at some later time.

Reply:
I agree and tried to address this with function documentation - but that's not ideal.

We don't have a good API pattern for optional function parameters. In e2epod, we ended up with lots of variants which just differ by the number and type of parameters. I think we should come up with something better, I just don't know yet what it is.

Reply:
Actually, e2epod has one such pattern: provide building blocks (get function, gomega matcher) and let the caller decide how to put them together.

In this case, the call could look like this:
```
gomega.Eventually(ctx, e2edaemonset.Get(f, ds)).[WithTimeout(...)].Should(e2edaemonset.HaveAllPodsReady)
```

Reply:
This is similar to the "return function for wait.Poll` approach from Antonio, except that it is for Gomega and more modular (getting the daemon set and assertions against it are separate).